src/factor_instances_provider/provider/factor_instances_cache.rs

```diff
@@ -0,0 +1,525 @@
+            for (agnostic_path, instances) in instances_by_agnostic_path {
+                let instances = instances.factor_instances();
+
+                if let Some(existing_for_path) = existing_for_factor.get_mut(&agnostic_path) {
+                    if let Some(fi) = instances
+                        .intersection(&existing_for_path.factor_instances())
+                        .next()
+                    {
+                        return Err(CommonError::CacheAlreadyContainsFactorInstance {
+                            derivation_path: fi.derivation_path(),
+                        });
+                    }
+
+                    if let Some(last) = existing_for_path.factor_instances().last() {
+                        if instances.first().unwrap().derivation_entity_base_index()
+                            != last.derivation_entity_base_index() + 1
+                        {
+                            warn!(
+                                "Non-contiguous indices, the index `{}` was skipped!",
+                                last.derivation_entity_base_index() + 1
+                            );
+                            skipped_an_index_resulting_in_non_contiguousness = true;
+                        }
+                    }
+                    existing_for_path.extend(instances);
+                } else {
+                    existing_for_factor.insert(agnostic_path, FactorInstances::from(instances));
+                }
+            }
+        } else {
+            self.map
+                .insert(*factor_source_id, instances_by_agnostic_path.0);
+        }
+
+        Ok(skipped_an_index_resulting_in_non_contiguousness)
+    }
+
+    /// Inserts all instance in `per_factor`.
+    pub fn insert_all(
+        &mut self,
+        per_factor: &IndexMap<FactorSourceIDFromHash, FactorInstances>,
+    ) -> Result<()> {
+        for (factor_source_id, instances) in per_factor {
+            _ = self.insert_for_factor(factor_source_id, instances)?;
+        }
+        Ok(())
+    }
+
+    /// Returns the MAX derivation entity index for the given `factor_source_id` and `index_agnostic_path`.
+    pub fn max_index_for(
+        &self,
+        factor_source_id: impl Borrow<FactorSourceIDFromHash>,
+        index_agnostic_path: impl Borrow<IndexAgnosticPath>,
+    ) -> Option<HDPathComponent> {
+        self.get_mono_factor(factor_source_id, index_agnostic_path)
+            .unwrap_or_default()
+            .factor_instances()
+            .into_iter()
+            .map(|fi| fi.derivation_entity_index())
+            .max()
+    }
+
+    /// Returns enough instances to satisfy the requested quantity for each factor source,
+    /// **OR LESS**, never more, and if less, it means we MUST derive more, and if we
+    /// must derive more, this function returns the quantities to derive for each factor source,
+    /// for each derivation preset, not only the originally requested one.
+    pub fn get_poly_factor_with_quantities(
```

Comment:
would be cool if anyone of you can clean this up, quite a fun challenge :) probably we can use the same code for all `DerivationPreset::all()` to produce some intermediary result, and then conditionally reduce into `CachedInstancesWithQuantitiesOutcome`...